mirascope/core/base/_utils/_get_dynamic_configuration.py

```diff
@@ -1,39 +1,54 @@
 from collections.abc import Awaitable, Callable
-from typing import Any, TypeVar, cast, overload
+from typing import Any, TypeVar, cast
+
+from typing_extensions import TypeIs
 
 from ..dynamic_config import BaseDynamicConfig
+from ..message_param import BaseMessageParam
 
 _BaseDynamicConfigT = TypeVar(
     "_BaseDynamicConfigT", bound=BaseDynamicConfig, covariant=True
 )
 
 
```

Comment:
why do we need both the sync and async methods here when we didn't before?

Reply:
In the previous get_dynamic_configuration
The result value of `fn` was just returned by calling `fn`, so there was no need to await inside get_dynamic_configuration. However, since it is necessary to check the return type of `fn` with `isinstance` in this PR, it is necessary to do await.
Therefore, async was prepared to enable await.

Reply:
oooh i see ok that makes sense.

do you think it's better to have these separate functions, or should we instead just define an async function internally and return calling that? for example:

```python
def get_dynamic_configuration(
    fn: Callable[..., _BaseDynamicConfigT | Awaitable[_BaseDynamicConfigT]],
    args: tuple[Any, ...],
    kwargs: dict[str, Any],
) -> _BaseDynamicConfigT | Awaitable[_BaseDynamicConfigT]:
    result = (
        fn._original_fn(*args, **kwargs)  # pyright: ignore [reportFunctionMemberAccess]
        if hasattr(fn, "_original_fn")
        else fn(*args, **kwargs)
    )
    if inspect.isawaitable(result):

        async def inner_async() -> Awaitable[_BaseDynamicConfigT]:
            result = await result
            if _is_message_param_list(result):
                return cast(_BaseDynamicConfigT, {"messages": result})
            return result

        return inner_async()

    if _is_message_param_list(result):
        return cast(_BaseDynamicConfigT, {"messages": result})
    return result
```

this way consuming the utility is easy and only ever requires a single import (rather than having to import both the sync and async versions), and then the overloads provide proper type hints as before

Reply:
I tried the your code. But, the `result` is not resolved in the scope 🤔 

```python

    async def inner_async() -> Awaitable[_BaseDynamicConfigT]:
>       result = await result
E       UnboundLocalError: cannot access local variable 'result' where it is not associated with a value

../mirascope/core/base/_utils/_get_dynamic_configuration.py:30: UnboundLocalError
```

Reply:
OK, It works fine
```python
            async_result = await result
            if _is_message_param_list(result):
                return cast(_BaseDynamicConfigT, {"messages": result})
            return async_result
```